src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -611,44 +625,204 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
 
             return async (target, httpContext) =>
             {
-                var bodyValue = defaultBodyValue;
-                var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
-                if (feature?.CanHaveBody == true)
-                {
-                    if (!httpContext.Request.HasJsonContentType())
-                    {
-                        Log.UnexpectedContentType(httpContext, httpContext.Request.ContentType, factoryContext.ThrowOnBadRequest);
-                        httpContext.Response.StatusCode = StatusCodes.Status415UnsupportedMediaType;
-                        return;
-                    }
-                    try
-                    {
-                        bodyValue = await httpContext.Request.ReadFromJsonAsync(bodyType);
-                    }
-                    catch (IOException ex)
-                    {
-                        Log.RequestBodyIOException(httpContext, ex);
-                        return;
-                    }
-                    catch (JsonException ex)
-                    {
+                (var bodyValue, var successful) = await TryReadBodyAsync(httpContext, factoryContext);
 
-                        Log.InvalidJsonRequestBody(httpContext, parameterTypeName, parameterName, ex, factoryContext.ThrowOnBadRequest);
-                        httpContext.Response.StatusCode = StatusCodes.Status400BadRequest;
-                        return;
-                    }
+                if (!successful)
+                {
+                    return;
                 }
+
                 await continuation(target, httpContext, bodyValue);
             };
         }
+
+        static async Task<(object? FormValue, bool Successful)> TryReadBodyAsync(
+            HttpContext httpContext,
+            FactoryContext factoryContext)
+        {
+            Debug.Assert(factoryContext.JsonRequestBodyParameter is not null, "TryReadFormAsync() should not be called if there is no JSON body parameter.");
+
+            var bodyType = factoryContext.JsonRequestBodyParameter.ParameterType;
+            var parameterTypeName = TypeNameHelper.GetTypeDisplayName(factoryContext.JsonRequestBodyParameter.ParameterType, fullName: false);
```

Comment:
Anything remotely expensive should be moved out of the per-request logic. We should be capturing the type name. I think this is also a good opportunity to capture only what we need from the FactoryContext.

Reply:
Yeah I did spot this one, and at one point I had this in the catch block as that was the only place the value was used. I moved it back as it was already like this, but meant to ask if it was intentional.

Will move it so it's only done if needed.

Reply:
Previously, we always captured the `parameterTypeName` during the call to `RequestDelegateFactory.Create(...)` rather than lazily doing the reflection every time there's an error during the request.